### PR TITLE
feat: Usenet improvements rollout (#119, closes #120, #121, #123-#129)

### LINKED ISSUE #120: Usenet: demote noisy Prowlarr Logger.info calls to :debug
Part of #119.

## Context

`lib/mydia/indexers/adapter/prowlarr.ex` logs at `:info` level for every search result during protocol detection. In production this floods logs with one line per release per search.

Lines involved:
- `prowlarr.ex:305` — logs raw result protocol fields
- `prowlarr.ex:316-318` — logs detection fallback path
- `prowlarr.ex:328` — logs skipped result reason

## Why

These are debugging breadcrumbs that should not be emitted at `:info` in normal operation. A single search of 5 indexers returning 50 results each writes 1000+ log lines today.

## Proposal

Change the four `Logger.info` calls in `prowlarr.ex` to `Logger.debug`. No behavior change, no schema change, no test change required.

## Acceptance criteria

- [ ] All four log calls demoted to `:debug`.
- [ ] Existing Prowlarr tests still pass.
- [ ] Manual search against a Prowlarr instance produces zero info-level output for routine result parsing.

## Effort

~4-line diff. Smallest item in the parent plan.

### LINKED ISSUE #122: Usenet: completion webhooks for SABnzbd & NZBGet
Part of #119.

## Context

Completion detection today is polling-only. `Mydia.Jobs.DownloadMonitor` (`lib/mydia/jobs/download_monitor.ex`, 301 LOC) polls every client periodically; when an NZB transitions to `:completed`, it enqueues `Mydia.Jobs.MediaImport`. Typical latency: 30s to 5min depending on the polling interval.

Both supported clients can push completion in real time:
- **SABnzbd** — \"Notification Script\" / external script can POST JSON.
- **NZBGet** — post-processing script (pp-script) can POST.

We have no endpoint to receive these.

## Why

- Reduces import latency from minutes to seconds.
- Removes a class of polling-related edge cases (history rotated before we polled — see existing 5-minute snooze fallback in `MediaImport`).
- Brings us closer to Sonarr/Radarr's UX.

## Proposal

1. **Endpoint**: `POST /api/usenet/completed` in a new `lib/mydia_web/controllers/usenet_webhook_controller.ex`. Accept both SABnzbd notification-script JSON and NZBGet pp-script payloads (small payload shape difference — branch on `User-Agent` or a `?client=` query param).
2. **Auth**: per-client signing secret. Add `webhook_secret :: string` column to `download_clients` (`lib/mydia/settings/download_client_config.ex`) with a migration. Generate at client creation. The secret is passed to the client as part of the script URL (`POST .../api/usenet/completed?secret=...`).
3. **Behavior**: look up download by client + nzb id (passed in payload), enqueue `MediaImport` immediately. Idempotent — if `MediaImport` is already running or done, no-op.
4. **Polling stays as fallback**: if no webhook hit within N minutes after completion-state observed via poll, the existing path still triggers import. Webhooks are purely additive.
5. **UI**: show the webhook URL in the client settings form so users can paste it into SABnzbd/NZBGet script config. Display alongside a script snippet they can copy.

## Acceptance criteria

- [ ] New endpoint accepts SABnzbd + NZBGet payloads, validates secret, returns 200.
- [ ] Migration adds `webhook_secret` to `download_clients`.
- [ ] `MediaImport` is idempotent w.r.t. duplicate triggers (polling + webhook racing).
- [ ] Admin UI shows the webhook URL + a copy-pasteable post-processing script per client type.
- [ ] Tests: valid webhook triggers import; invalid secret returns 401; duplicate trigger no-ops.

## Effort

Medium. Touches webhook controller, migration, two adapters' setup docs, admin UI, `MediaImport` idempotency.

### PR DESCRIPTION
**Pull request by arsfeld**
## Summary

Ships 9 of the 10 open Usenet improvement issues tracked under #119. (#122 was the completion-webhook issue — see "Why no webhook" below; closed separately.)

**Closes:** #119, #120, #121, #123, #124, #125, #126, #127, #128, #129
**Does not close:** #122 (closed separately as a design rejection)

### What ships per issue

- **#120** — Demoted noisy Prowlarr `Logger.info` calls to `:debug`.
- **#121** — `SearchResult.usenet_date`; per-indexer `IndexerConfig.min_post_age_minutes` (nullable); search jobs filter too-fresh NZB results via `Indexers.reject_too_fresh_nzbs/1` alongside `reject_blacklisted/2`. Admin form exposes the field.
- **#123** — `release_blacklist` table; `Blacklists.add/4`, `blacklisted?/2`, `reject_blacklisted/2` (single batched query); `DownloadMonitor.handle_failure/1` writes a blacklist row on `:error` (non-blocking via try/rescue); `TvShowSearch` and `MovieSearch` filter blacklisted releases pre-ranking; `/admin/release-blacklist` admin LiveView; `BlacklistCleanup` Oban worker on daily cron.
- **#124** — `download_client_configs.categories :map` resolved by `Download.content_type` at submission time; backwards-compat fallback to legacy `category` string.
- **#125** — `SearchResult.nzb_completion` and `nzb_grabs` replace the misuse of `seeders`/`leechers` for NZB results; ReleaseRanker scoring branches on `download_protocol`.
- **#126** — `last_progress_at` + `last_known_bytes` columns; `StallDetector` pure module with strict-`>` grace boundary and byte-regression tolerance; `DownloadMonitor` flags stalled downloads to `import_failed_at` after `incomplete_grace_minutes`; downloads LiveView surfaces a "Stalled" badge.
- **#127** — `TorrentStatus` → `DownloadStatus` rename; new `Mydia.Downloads.Client.Helpers` (`parse_size_mb_to_bytes/1`, `parse_size_bytes/1`, `parse_timestamp_unix/1`, `priority_profile/1`). All 6 adapters use the shared struct + helpers.
- **#128** — Frozen JSON fixtures under `test/support/fixtures/{sabnzbd,nzbget,prowlarr}/`; table-driven state-mapping tests covering every documented status; SABnzbd ETA edge cases, NZBGet computed ETA, Prowlarr protocol-detection priority. **+140 deterministic tests, no network.**
- **#129** — `Mydia.Downloads.Priority` 5-tier taxonomy (`:verylow..:veryhigh`); `download_client_configs.priority_profile :map`; all 6 adapters resolve priority via profile with backwards-compat default mapping.

### Adaptive polling (replaces the would-be webhook)

`DownloadMonitor` (2-min cron) now seeds a chain of fast follow-up jobs when active downloads exist:

- Each cron tick inspects the active-download set; if non-empty, schedules itself 15s out with an incremented `fast_chain_position` flag.
- The chain runs up to 7 hops (~105s) and stops the moment no active downloads remain.
- `DownloadMonitor`'s own `unique: [period: 120, states: [:available, :scheduled]]` prevents the cron and a chain from stacking up.
- Effect: a completed download lands in the library ~15s after the client says so, with **zero operator configuration**. Idle overhead is unchanged.

### Why no webhook (closes #122 explicitly NOT addressed)

The original plan included a `POST /api/webhooks/usenet/:client_id` endpoint with per-client signed-secret auth + admin UI rendering the URL + script snippets to paste into SABnzbd/NZBGet. **That whole shape is wrong for a self-hosted app:**

- We don't know what hostname/port the operator's Mydia is reachable at from their downloader's network (Docker compose service name? `host.docker.internal`? LAN IP? Reverse-proxy hostname? HTTPS or not?).
- `Endpoint.url()` returns the value *Mydia* believes, not the value any *other* box on the network believes.
- Even if we displayed a "best guess" URL, the operator would silently paste a wrong one, the integration would silently not fire, and the only feedback is "feature doesn't work."
- Asking operators to install a Python or bash script in their downloader's config folder makes the integration's correctness dependent on the operator getting deployment-topology details right that we can't help them with.

Adaptive polling fills the latency gap without crossing that boundary. #122 closed as a design rejection in favor of the polling fix.

### Test plan

- [x] `./dev mix test test/mydia/ test/mydia_web/` — 4250 tests, 0 failures, 34 skipped
- [x] `./dev mix ecto.migrate` runs cleanly across all 3 migrations (down from 4 — webhook backfill removed); rollback verified clean
- [x] Adapter parsing tests against frozen fixtures (U9): 140 tests across SABnzbd/NZBGet/Prowlarr
- [x] Integration test for SABnzbd stalled-detection happy path (U4)

### Known limitations / follow-up items

- **Stringly-typed `failure_reason`** (`"stalled"`, `"client_reported_failure"`) — should become a module-attribute constant set.
- **Single content-type source of truth** — `"movie"`/`"tv"`/`"music"` keys appear in 3 places.
- **qBittorrent priority is a no-op stub** — qBittorrent's `/torrents/add` endpoint doesn't accept priority; lookup runs and logs at `:debug`. Follow-up: emit a `topPrio` call.
- **mock-sabnzbd / mock-nzbget E2E containers** were promised by an earlier plan but never shipped (verified). Tests use `Bypass`/`Phoenix.ConnTest` instead.
- Numerous P3 coverage gaps and minor doc/code mismatches — recorded in the code-review run artifact for incremental follow-up.

## Post-Deploy Monitoring & Validation

- **Migrations to verify:**
  ```sql
  PRAGMA table_info(indexer_configs);              -- min_post_age_minutes
  PRAGMA table_info(download_client_configs);      -- categories, priority_profile, incomplete_grace_minutes
  PRAGMA table_info(downloads);                    -- last_progress_at, last_known_bytes
  SELECT count(*) FROM release_blacklist;          -- new table
  ```
- **Log signals to watch:** `"Filtered too-fresh NZB"`, `"Rejected blacklisted release"`, `"Release blacklisted after failure"`, `"Download stalled — no progress within grace window"`. Spikes of `"Failed to blacklist release"` would indicate the `record_blacklist_entry/2` rescue branch firing → investigate.
- **Adaptive polling cost check:** count `DownloadMonitor` job inserts over 24h. Active hours should see ~10/min during active downloads; idle should stay at the cron baseline (0.5/min).
- **Validation window:** 24h for the import path; 1 week for search/ranking changes.
- **Owner:** Alexandre Rosenfeld.

🤖 [Compound Engineered](https://every.to/compound-engineering) with [Claude Code](https://claude.com/claude-code) using Opus 4.7.

### CODE REVIEW

File: lib/mydia/indexers.ex [hunk 1]
```diff
@@ -437,11 +437,79 @@ defmodule Mydia.Indexers do
...
+       )
+       when is_binary(indexer) do
+    case Map.get(configs_by_name, indexer) do
+      minutes when is_integer(minutes) and minutes > 0 ->
+        DateTime.diff(now, posted, :second) < minutes * 60
+
+      _ ->
+        false
+    end
+  end
```
[the remark this answers is not in the document]
Comment (arsfeld) [reply lib/mydia/indexers.ex]: Good catch — fixed in 2861c69.

You're right that `result.indexer` is the upstream label (e.g. `"DOGnzb"`) for relay-style indexers like Prowlarr/Jackett, so keying the lookup off it against a map of `IndexerConfig.name` was a silent no-op for the common case. Instead of threading `indexer_config_id`/`indexer_config_name` onto `SearchResult` (which would touch every adapter), the filter now runs inside `Indexers.search_with_metrics/3` where the originating `IndexerConfig` is already in scope. Each batch of results is filtered against its own config's `min_post_age_minutes` before being merged. The relay vs. configured-name mismatch becomes a non-issue because we never key the lookup off `result.indexer`.

Removed the global-map `Indexers.reject_too_fresh_nzbs/2` and the now-redundant calls in `TvShowSearch`/`MovieSearch` (left a breadcrumb comment at the old call sites pointing at the new home). Cardigann definitions, which don't carry the setting, pass through unchanged.

File: lib/mydia_web/controllers/api/webhook/usenet_controller.ex [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+              client_name: client_config.name,
+              remote_id: remote_id
+            )
+
+            send_resp(conn, 404, "")
+
+          download ->
+            enqueue_media_import(download)
+            send_resp(conn, 200, "")
+        end
```
[the remark this answers is not in the document]
Comment (arsfeld) [reply lib/mydia_web/controllers/api/webhook/usenet_controller.ex]: Moot — the entire webhook receiver was removed in 6cfaaf1 ("refactor(downloads): replace webhook receiver with adaptive polling"). Mydia is self-hosted and the deployment-topology assumptions baked into the webhook design (operators figuring out what URL their Mydia is reachable at from their SAB/NZBGet box and pasting a Python/bash script into the downloader's config) push deployment decisions onto users that we can't help them get right. Latency gap is filled by an adaptive-polling chain in `DownloadMonitor` that runs 15s-spaced follow-ups while active downloads exist and self-terminates when idle.

#122 (the webhook issue) closed as a design rejection. The 200-on-enqueue-failure inconsistency you flagged was real for the code that existed at review time — just preempted by the bigger architectural change.
